FEAT - Basic GeneralizedLinearEstimatorCV

floriankozikowski opened the pull request:
## Context of the PR


Implements a cross-validation wrapper for `GeneralizedLinearEstimator` along with a standalone example script. Provides hyperparameter grids over `alpha` and optional `l1_ratio`, custom k-fold splitting, warm starts, scoring, and final full-data refit (closes Issue  #308 ).


## Contributions of the PR

<!-- List the contribution of the PR -->

- class inherits from GeneralizedLinearEstimator, CV inspired by Celer
- full pass over l1_ratio (if available) and alpha grids
- warm_start: carry forward each fold’s solution (w) as the initial w_start for the next alpha
- score & select: compute MSE (or user-provided scorer) per fold, track the mean loss in mse_path, and pick the (alpha_, l1_ratio_) with lowest average loss
- final refit: once the best hyperparameters are found, update self.penalty in place and call super().fit(X, y) to train the returned model on the entire dataset
- plot_generalized_linear_estimator.cv provides a simple example file

### Checks before merging PR

- [x] added documentation for any new feature
- [x] added unit tests
- [x] edited the [what's new](../doc/changes/whats_new.rst) (if applicable)

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+            return -float(scorer._score_func(y_true, y_pred, **scorer._kwargs))
+
+        if isinstance(self.datafit, (Logistic, QuadraticSVC)):
+            return -float(np.mean(y_true == (y_pred > 0)))
```
and also, why comparing to y_pred > 0 and not y_pred?

floriankozikowski replied:
float I honestly just used for convenience, but I could also call .item() if you prefer that. 
For y_pred, you are right I will change it, its redundant as self.predict(X) already returns class labels. 

Otherwise I could just use sklearn.metrics.accuracy_score and mean_squared_error if thats better

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+        """Fit the model using cross-validation."""
+        if not hasattr(self.penalty, "alpha"):
+            raise ValueError("'penalty' must expose an 'alpha' hyper-parameter.")
+        X, y = np.asarray(X), np.asarray(y)
```
why this ? this breaks sparsity if user passes sparse scipy matrix

floriankozikowski replied:
true, I remove np.asarray(X), but I keep y since targets are always dense, or am I missing something?

mathurinm replied:
all preprocessing of design matrix and target is done in GeneralizedLinearEstimator.fit. You need to leverage tthis and not do anything here

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+        if not hasattr(self.penalty, "alpha"):
+            raise ValueError("'penalty' must expose an 'alpha' hyper-parameter.")
+        X, y = np.asarray(X), np.asarray(y)
+        n, p = X.shape
```
we use n_samples, n_features as notation

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+            alphas = np.sort(self.alphas)[::-1]
+        else:
+            alpha_max = np.max(np.abs(X.T @ y)) / n
+            alphas = np.logspace(
```
use np.geomspace to avoid get the result directly

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+        l1_ratios = [1.] if not has_l1_ratio else np.atleast_1d(
+            self.l1_ratio if self.l1_ratio is not None else [1.])
+
+        folds = list(_kfold_split(n, self.cv, rng))
```
you probably can avoid building the full list, that the goal of an enumerator. not crucial here

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+        folds = list(_kfold_split(n, self.cv, rng))
+        n_folds = len(folds)
+
+        mse_path = np.empty((len(l1_ratios), len(alphas), n_folds))
```
if l1_ratio is None we want a 1d array directly, not 2D

floriankozikowski replied:
understood, but you mean 2D and 3D array, right? As we still have multiple alphas and multiple folds or no?

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+                          if k not in ("alpha", "l1_ratio")}
+            if has_l1_ratio:
+                pen_kwargs['l1_ratio'] = l1
+            pen = type(self.penalty)(alpha=alpha, **pen_kwargs)
```
you could just do `self.penalty.alpha = alpha`, no ?

floriankozikowski replied:
I thought since we run the folds in parallel, changing self.penalty.alpha directly could cause issues if multiple jobs modify it at the same time, so I created a new penalty for each fold

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+            pen = type(self.penalty)(alpha=alpha, **pen_kwargs)
+
+            kw = dict(X=X[train], y=y[train], datafit=self.datafit, penalty=pen)
+            if 'w' in self.solver.solve.__code__.co_varnames:
```
what does this do ? this seems overly complex

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+            if 'w' in self.solver.solve.__code__.co_varnames:
+                kw['w'] = w_start
+            w = self.solver.solve(**kw)
+            w = w[0] if isinstance(w, tuple) else w
```
which solvers return a tuple ?

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+
+            coef, intercept = (w[:p], w[p]) if w.size == p + 1 else (w, 0.0)
+
+            y_pred = X[test] @ coef + intercept
```
for logistic regression the prediction is not this, it's the sign of this vector

you can delegate everything to a GeneralizedEstimator (creating it by passing datafit, penalty, solver, then changing the alpha attribute and using warm_start=True)

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+        self.mse_path_ = mse_path
+        return self
+
+    def predict(self, X):
```
delegate to an instance of GeneralizedLinearEstimator instead

mathurinm commented on skglm/penalties/generalized_linear_cv.py:
```diff
@@ -0,0 +1,147 @@
+        super().fit(X, y)
+        self.alphas_ = alphas
+        self.mse_path_ = mse_path
+        return self
```
Look at how LassoCV works: after fitting, you can directly call it to predict, because it stores the best Lasso. you need to do the same, storing the best model coefficient, the best  `best_alpha_ ` and recreating a GeneralizedLienarModel at the end, that fits on the whole dataset X:

look at how it's done here : 
https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/linear_model/_coordinate_descent.py#L1882

mathurinm commented on skglm/cv.py:
```diff
@@ -0,0 +1,139 @@
+from skglm.estimators import GeneralizedLinearEstimator
+
+
+def _kfold_split(n_samples, k, rng):
```
can you rely on sklearn KFold() instead ? It uses stratified kfold by default and should yield shorter code (less error prone)

mathurinm commented on skglm/cv.py:
```diff
@@ -0,0 +1,132 @@
+            est.solver.warm_start = True
+            est.fit(X[train], y[train])
+            y_pred = est.predict(X[test])
+            return est.coef_, est.intercept_, self._score(y[test], y_pred)
```
Would it make more sense to delegate to est.score here?

mathurinm commented:
@floko can you provide some timings on datasets of reasonable sizes ?


We should handle warm start better, by storing `K` estimators (one for each fold) and using warm start  on them (so that on fold `k`, when fitting for one value of `alpha`, we use warm start with the previous value of `alpha` on the same fold.

Right now warm start is not used since we recreate an estimator at each alpha.

mathurinm commented on skglm/tests/test_cv.py:
```diff
@@ -0,0 +1,37 @@
+    """Test GeneralizedLinearEstimatorCV matches sklearn GridSearchCV for ElasticNet."""
+    X, y = make_regression(n_samples=100, n_features=20, noise=0.1, random_state=seed)
+
+    alphas = np.array([0.001, 0.01, 0.1, 1.0])
```
picking absolute values for alphas is dangerous ; scale them by `\Vert X.T @ y \Vert_\infty / n` which is the smallest alpha for which the solution is 0

floriankozikowski commented:
@Badr-MOUFAD good catch, I removed it from the PR description (initially I had it like this, but then changed it at some point). Plus, I made the unit test shorter now. Lets me know if thats fine.
@mathurinm initially you added a part where you also test for alpha = 1e-4. For now I removed it as it was in the debugging sections of the script. But if you want it back let me know. When I tried it out the pytest still worked, but was much slower and I get convergence warnings.

Badr-MOUFAD commented on skglm/tests/test_cv.py:
```diff
@@ -0,0 +1,42 @@
+        alphas=alphas, l1_ratio=l1_ratios, cv=5, random_state=seed, n_jobs=1
+    ).fit(X, y)
+
+    assert sklearn_model.best_params_['alpha'] == skglm_model.alpha_
+    assert sklearn_model.best_params_['l1_ratio'] == skglm_model.l1_ratio_
```
we better use [np assert equal](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_equal.html), it provides better feedback

Badr-MOUFAD commented on skglm/cv.py:
```diff
@@ -0,0 +1,181 @@
+                    for k, (tr, te) in enumerate(split_iter)
+                )
+
+                for k, (coef_fold, intercept_fold, loss_fold) in \
+                        enumerate(fold_results):
+                    warm_start[k] = (coef_fold, intercept_fold)
+                    scores_path[idx_ratio, idx_alpha, k] = loss_fold
```
to fit in one line

```suggestion
                for k, (coef_k, intercept_k, loss_k) in enumerate(fold_results):
                    warm_start[k] = (coef_k, intercept_k)
                    scores_path[idx_ratio, idx_alpha, k] = loss_k
```

Badr-MOUFAD commented on skglm/cv.py:
```diff
@@ -0,0 +1,181 @@
+        self.eps = eps
+        self.n_alphas = n_alphas
+
+    def _score(self, y_true, y_pred):
+        """Compute the performance score (higher is better)."""
+        if isinstance(self.datafit, (Logistic, QuadraticSVC)):
+            return float(np.mean(y_true == y_pred))
+        return -float(np.mean((y_true - y_pred) ** 2))
```
I'm +1 with using sklearn `accuracy_score` and `mean_squared_error`
this forces us to stay consistent with the sklearn API, as we are drop in replacement of it

Badr-MOUFAD reviewed (approved):
LGMT @floriankozikowski, thanks for this amazing job 💪 

just minor remarks before merging
